### DISCUSSION TITLE
Jails on boot drive

### DISCUSSION THREAD

**Discussion by spusuf** [Jip-Hop/jailmaker#56] [General]
TrueNAS is notorious for disallowing any use of the boot drive, leaving 100gb unused on a basic 120gb drive.
Would it be feasible to use jails in this space?
Are there any drawbacks or limitations?
Does the install script work from the boot drive (root of filesystem) as opposed to within a dataset?

I am planning to test this myself, but wanted a thread for anyone with similar testing. I'm not going to do this on my main server, just waiting for the time to reset my second server.

**Comment by Jip-Hop** [comment]
Why would you want to do that? The root filesystem is completely overwritten when you upgrade to the next (minor) release. All your jail files and config would be lost. It's kind of the whole point of jailmaker to prevent / work around this. Otherwise you could just use systemd-nspawn directly...

**Reply by spusuf** [reply]
if that can be overcome through backups then it would be a much better use of the space then leaving the boot drive completely empty.

**Reply by Jip-Hop** [reply]
In that case just play with systemd-nspawn directly. It should be pre-installed (again). Jailmaker doesn't and will not support this.

**Reply by Lockszmith-GH** [reply]
Totally agree with Jip-Hop, if your stung up about the unused disk space, repartition the drive. That at least survives an upgrade.

**Reply by Lockszmith-GH** [reply]
@spusuf check out [this](https://gist.github.com/gangefors/2029e26501601a99c501599f5b100aa6) and/or this [reddit post](https://www.reddit.com/r/truenas/comments/lgf75w/scalehowto_split_ssd_during_installation/)

**Reply by spusuf** [reply]
I didnt see this until now, but partitioning during boot can (and did for the few times I tried) lead to non boot. I'd much rather have a config file on a persistent location and have jailmaker recreate the jail if necessary (a la docker compose). But I understand this is out of scope. 

This is one of the main reasons I have switched to Debian. Losing an entire drive for such little operations and storage that it could be done by emmc on a small appliance/low end server with only 4x sata + occupied PCIe, is entirely unreasonable.

**Reply by Lockszmith-GH** [reply]
I use boot drive partitioning, but not for extra storage, mostly for performance.
I would agree that IX got the right idea of having the boot drives disposable, I just tweak them because small SSD drives are no longer a feasable thing to buy, and so not utilizing the rest of the drives seems like a waste.
So I'm agree with you @spusuf, I'd rather not put any precious storage on the boot drives either.

I can share my experience though of this working for over a year on multiple systems, surviving IX upgrades. Luckily never had a boot drive hardware failure or corruption, but I did test failover which gave me both peace of mind + I got the performance boost for the mechanical drives, which is what I wanted.

But my point is, repartitioning the boot drive correctly survives IX system upgrades, and does not affect the boot process.

While I understand why IX doesn't offer this out of the box, This has become my standard method of operation with TrueNAS SCALE.
It's a hassle to setup, but once it's setup - no further tinkering is required.

<details><summary><h2>My setup, in case you are interested</h2> (click to expand)</summary>

I did not come up with this design, stumbled on it in the old TrueNAS forums a while back.

> NOTE: I've simplified the output to use short device name, but I recommend using PartUUIDs instead, which is what I do.

My **boot-pool** pool is a mirror of 2x 46GB partitions (_sdc3_, _sdd3_) out of 2x 256 GB SSD drives (_sdc_, _sdd_).
My **tank** pool is a mirror of 2x 1TB mechanical drives (_sda_, _sdb_).

Each of the boot drives are partitioned the same way, contianing an EFI boot parition (_sdc2_, _sdd2_), a swap partitoin (_sdc4_, _sdd4_).
The remainder of the SSD boot drivers (~170Gb) are assigned to the tank pool as [L2ARC cache](https://www.truenas.com/docs/references/l2arc/) (_sdd5_) and [slog for ZIL](https://www.truenas.com/docs/references/slog/) (_sdc5_), improving the performance of ZFS on the mechanical drives.

```console
# zpool status
  pool: boot-pool
 state: ONLINE
...
config:
        NAME        STATE     READ WRITE CKSUM
        boot-pool   ONLINE       0     0     0
          mirror-0  ONLINE       0     0     0
            sdc3    ONLINE       0     0     0
            sdd3    ONLINE       0     0     0
...
  pool: tank
 state: ONLINE
...
config:
        NAME        STATE     READ WRITE CKSUM
        tank        ONLINE       0     0     0
          mirror-0  ONLINE       0     0     0
            sdb2    ONLINE       0     0     0
            sda2    ONLINE       0     0     0
        logs
          sdc5      ONLINE       0     0     0
        cache
          sdd5      ONLINE       0     0     0
          
# lsblk --output=NAME,SIZE,TYPE,FSTYPE,LABEL
NAME          SIZE TYPE  FSTYPE            LABEL
sda         931.5G disk
├─sda1          2G part
└─sda2      929.5G part  zfs_member        tank
sdb         953.9G disk
├─sdb1          2G part
└─sdb2      951.9G part  zfs_member        tank
sdc         232.9G disk
├─sdc1          1M part
├─sdc2          2G part  vfat              EFI
├─sdc3         46G part  zfs_member        boot-pool
├─sdc4         16G part  linux_raid_member swap0
│ └─md127      16G raid1
│   └─md127    16G crypt swap
└─sdc5      168.9G part  zfs_member        tank
sdd         232.9G disk
├─sdd1          1M part
├─sdd2          2G part  vfat              EFI
├─sdd3         46G part  zfs_member        boot-pool
├─sdd4         16G part  linux_raid_member swap0
│ └─md127      16G raid1
│   └─md127    16G crypt swap
└─sdd5      168.9G part  zfs_member        tank
zd0            32G disk
zd16           32G disk
```
</details>
